src/OpenTelemetry/Metrics/Aggregator/LastValueAggregator.cs

```diff
@@ -0,0 +1,80 @@
+// <copyright file="LastValueAggregator.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Linq;
+
+namespace OpenTelemetry.Metrics
+{
+    public class LastValueAggregator : Aggregator
+    {
+        private readonly KeyValuePair<string, object>[] emptyKVP = new KeyValuePair<string, object>[0];
+        private readonly Instrument instrument;
+        private readonly string[] names;
+        private readonly object[] values;
+
+        private IDataPoint lastValue;
+        private int count = 0;
+
+        public LastValueAggregator(Instrument instrument, string[] names, object[] values)
+        {
+            if (names.Length != values.Length)
+            {
+                throw new ArgumentException("Length of names[] and values[] must match.");
+            }
+
+            this.instrument = instrument;
+            this.names = names;
+            this.values = values;
+        }
+
+        public override void Update<T>(DateTimeOffset dt, T value)
+            where T : struct
+        {
+            this.count++;
+            this.lastValue = new DataPoint<T>(dt, value, this.emptyKVP);
```

Comment:
@victlu FYI I was curious why this wasn't showing an allocation in the benchmarks so I got your branch and poked around a bit. The reason is this test in `GetDefaultAggregator`:

```csharp
            if (!this.instrument.IsObservable)
            {
                return new Aggregator[]
                {
                    new SumAggregator(this.instrument, seqKey, seqVal),
                };
            }
```

The benchmarks at the moment only cover `Counter<int>` which falls into that. When I commented it out so it uses `LastValueAggregator` I get these results:

|                    Method | WithSDK |       Mean |     Error |     StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |-------- |-----------:|----------:|-----------:|-------:|------:|------:|----------:|
|            CounterHotPath |   False |   3.978 ns | 0.0499 ns |  0.0466 ns |      - |     - |     - |         - |
| CounterWith1LabelsHotPath |   False |  14.309 ns | 0.1021 ns |  0.0797 ns |      - |     - |     - |         - |
| CounterWith3LabelsHotPath |   False |  37.375 ns | 0.7576 ns |  1.2657 ns |      - |     - |     - |         - |
| CounterWith5LabelsHotPath |   False |  27.288 ns | 0.2998 ns |  0.2504 ns | 0.0124 |     - |     - |     104 B |
|            CounterHotPath |    True | 112.516 ns | 1.7106 ns |  1.4284 ns | 0.0057 |     - |     - |      48 B |
| CounterWith1LabelsHotPath |    True | 191.771 ns | 1.3932 ns |  1.1633 ns | 0.0057 |     - |     - |      48 B |
| CounterWith3LabelsHotPath |    True | 554.198 ns | 9.5792 ns | 12.1146 ns | 0.0057 |     - |     - |      48 B |
| CounterWith5LabelsHotPath |    True | 661.714 ns | 5.1245 ns |  4.5427 ns | 0.0334 |     - |     - |     280 B |

The reason is `DataPoint` has to be boxed up and copied to heap to be set as `IDataPoint`. I don't have an immediate solution to be more helpful sorry. It kind of depends on how we're going to go from DataPoint to value/Metric?

Reply:
I got this!